### PR TITLE
[SPARK-12913] [SQL] Improve performance of stat functions

### PR DESCRIPTION
**Pull request by davies**
As benchmarked and discussed here: https://github.com/apache/spark/pull/10786/files#r50038294, benefits from codegen, the declarative aggregate function could be much faster than imperative one.

### CODE REVIEW

**Comment by davies** [comment]
cc @mengxr

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 1]
```diff
@@ -109,7 +109,7 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
    * Update the central moments buffer.
    */
   override def update(buffer: MutableRow, input: InternalRow): Unit = {
-    val v = Cast(child, DoubleType).eval(input)
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Creating a Cast() here is very expensive

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 2]
```diff
@@ -125,19 +125,15 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
-      if (momentOrder >= 2) {
-        m2 = buffer.getDouble(secondMomentOffset)
-        m2 += delta * (delta - deltaN)
-        buffer.setDouble(secondMomentOffset, m2)
-      }
+      m2 = buffer.getDouble(secondMomentOffset)
+      m2 += delta * (delta - deltaN)
+      buffer.setDouble(secondMomentOffset, m2)
 
-      if (momentOrder >= 3) {
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Those `if` branches are important to save computation for low-order statistics. Even we won't use `CentralMomentAgg` for second-order statistics, it is still good to keep them.

**Comment by mengxr** [comment]
@davies Did you get a chance to test whole-stage codegen with higher-order statistics like skewness? If it works, the cleanest solution would be changing `CentralMomentAgg` to declarative and then make all existing univariate summary statistics call it.

**Comment by mengxr** [comment]
@davies side note: The JIRA number is wrong.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/WindowQuerySuite.scala [hunk 3]
```diff
@@ -133,44 +133,44 @@ class WindowQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleto
         Row("Manufacturer#1", "almond antique burnished rose metallic", 2, 258.10677784349247, 258.10677784349247, 34, 66619.10876874997, 0.811328754177887, 2801.7074999999995),
         Row("Manufacturer#1", "almond antique burnished rose metallic", 2, 273.70217881648085, 273.70217881648085, 2, 74912.88268888886, 1.0, 4128.782222222221),
         Row("Manufacturer#1", "almond antique burnished rose metallic", 2, 273.70217881648085, 273.70217881648085, 34, 74912.88268888886, 1.0, 4128.782222222221),
-        Row("Manufacturer#1", "almond antique chartreuse lavender yellow", 34, 230.9015158547037, 230.9015158547037, 2, 53315.510023999974, 0.6956393773976641, 2210.7864),
```
Comment (davies) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/WindowQuerySuite.scala]: The new implementation of Corr/Covar have better accuracy, so updated the tests to match that.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 4]
```diff
@@ -790,7 +790,7 @@ abstract class AggregationQuerySuite extends QueryTest with SQLTestUtils with Te
         """
           |SELECT corr(b, c) FROM covar_tab WHERE a = 3
         """.stripMargin),
-      Row(null) :: Nil)
+      Row(Double.NaN) :: Nil)
```
Comment (davies) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: This is changed to match the behavior as stddev_samp and others.

**Comment by davies** [comment]
@nongli @rxin Could you review this one?

**Comment by mengxr** [comment]
I'm making a pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 5]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
 
   override def dataType: DataType = DoubleType
 
-  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType)
+  override def inputTypes: Seq[AbstractDataType] = Seq(DoubleType)
+
+  protected val count = AttributeReference("count", DoubleType, nullable = false)()
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Shall we call it `n`? Otherwise, we have both `count` and `n` in the code. It is not clear that `n` means `newCount`. If we rename it, we can use `n` and `newN`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 6]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
+  protected val avg = AttributeReference("avg", DoubleType, nullable = false)()
+  protected val m2 = AttributeReference("m2", DoubleType, nullable = false)()
+  protected val m3 = AttributeReference("m3", DoubleType, nullable = false)()
+  protected val m4 = AttributeReference("m4", DoubleType, nullable = false)()
+
+  override val aggBufferAttributes = Seq(count, avg, m2, m3, m4).take(momentOrder + 1)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Shall we make `.take(momentOrder + 1)` a private method (`trimHigherOrder`?) and have it documented?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 7]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
+  override lazy val updateExpressions: Seq[Expression] = {
+    val n = count + Literal(1.0)
+    val delta = child - avg
+    val deltaN = delta / n
+    val newAvg = avg + deltaN
+    val newM2 = m2 + delta * (delta - deltaN)
+
+    val delta2 = delta * delta
+    val deltaN2 = deltaN * deltaN
+    val newM3 = m3 - Literal(3.0) * deltaN * newM2 + delta * (delta2 - deltaN2)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: If my understanding is correct, there exists some unnecessary expression calculation when we don't need 3rd- or 4th-order statistics. But this is a one-time overhead because `updateExpression` is only called once before codegen. If this is correct, it would be helpful to leave a comment here to explain why we don't need `if` branches.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 8]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
+
+    val newM4 = m4 - Literal(4.0) * deltaN * newM3 - Literal(6.0) * deltaN2 * newM2 +
+      delta * (delta * delta2 - deltaN * deltaN2)
+
+    if (child.nullable) {
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: It might be easier to understand if we write

~~~scala
trimHigherOrder(
  If(IsNull(child),
    Seq(count, avg, m2, m3, m4),
    Seq(n, newAvg, newM2, newM3, newM4)))
~~~

Please let me know if it doesn't work. Btw, I guess `if (child.nullable)` is automatically checked by `IsNull` (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala#L74). Then we can remove that check to simplify the code here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 9]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
+      Seq(
+        /* count = */ If(IsNull(child), count, n),
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: The comment `/* count = */` seems unnecessary to me because it is clear from the variable name.

Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Please also remove comments when the variable names provide sufficient info.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 10]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
-  override def checkInputDataTypes(): TypeCheckResult =
-    TypeUtils.checkForNumericExpr(child.dataType, s"function $prettyName")
+  override lazy val mergeExpressions: Seq[Expression] = {
 
-  override def aggBufferSchema: StructType = StructType.fromAttributes(aggBufferAttributes)
+    val n1 = count.left
+    val n2 = count.right
+    val n = n1 + n2
+    val delta = avg.right - avg.left
+    val deltaN = If(EqualTo(n, Literal(0.0)), Literal(0.0), delta / n)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Shall we use `===` to replace `EqualTo` to make the code more readable?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 11]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
+    val newAvg = avg.left + deltaN * n2
 
-  /**
-   * Size of aggregation buffer.
-   */
-  private[this] val bufferSize = 5
+    // higher order moments computed according to:
+    // https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Higher-order_statistics
+    val newM2 = m2.left + m2.right + delta * deltaN * n1 * n2
+    val newM3 = if (momentOrder >= 3) {
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: The `if` branch is not necessary because we will trim the result at the end. Shall we follow the approach in `updateExpression` and remove it? It is still useful to leave a comment here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 12]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
-  private[this] var deltaN = 0.0
-  private[this] var delta2 = 0.0
-  private[this] var deltaN2 = 0.0
-  private[this] var n = 0.0
-  private[this] var mean = 0.0
-  private[this] var m2 = 0.0
-  private[this] var m3 = 0.0
-  private[this] var m4 = 0.0
+// Compute the population standard deviation of a column
+case class StddevPop(child: Expression) extends CentralMomentAgg(child) {
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Btw, `StddevPop` is simply `Sqrt(VariancePop)` (and `StddevVSamp = Sqrt(VarianceSamp)`). I'm not sure whether it can help simplify the code here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 13]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
-   * Initialize all moments to zero.
-   */
-  override def initialize(buffer: MutableRow): Unit = {
-    for (aggIndex <- 0 until bufferSize) {
-      buffer.setDouble(mutableAggBufferOffset + aggIndex, 0.0)
-    }
+  override protected def momentOrder = 2
+
+  override val evaluateExpression: Expression = {
+    If(EqualTo(count, Literal(0.0)), Literal.create(null, DoubleType),
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Using `===` instead?

~~~scala
If(count === Literal(0.0), Literal.create(null, DoubleType), Sqrt(m2 / count))
~~~

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 14]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
+  override def prettyName: String = "stddev_pop"
+}
+
+// Compute the sample standard deviation of a column
+case class StddevSamp(child: Expression) extends CentralMomentAgg(child) {
+
+  override protected def momentOrder = 2
+
+  override val evaluateExpression: Expression = {
+    If(EqualTo(count, Literal(0.0)), Literal.create(null, DoubleType),
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: ditto on `EqualTo` => `===`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala [hunk 15]
```diff
@@ -29,165 +28,95 @@ import org.apache.spark.sql.types._
...
+    /* count = */ Literal(0.0),
+    /* xAvg = */ Literal(0.0),
+    /* yAvg = */ Literal(0.0),
+    /* ck = */ Literal(0.0),
+    /* xMk = */ Literal(0.0),
+    /* yMk = */ Literal(0.0)
+  )
+
+  override lazy val updateExpressions: Seq[Expression] = {
+    val n = count + Literal(1.0)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala]: Same argument on `count` and `n` usage. Calling them `n` and `newN` might be better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala [hunk 16]
```diff
@@ -29,165 +28,95 @@ import org.apache.spark.sql.types._
...
+    val dx = x - xAvg
+    val dxN = dx / n
+    val dy = y - yAvg
+    val dyN = dy / n
+    val newXAvg = xAvg + dxN
+    val newYAvg = yAvg + dyN
+    val newCk = ck + dx * (dy - dyN)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala]: This is equivalent to the original update rule. However, for reference purpose, we shouldn't change it. So users can easily map the formula on the wikipedia page to the implementation here. Shall we change it back to `y - newYAvg`? Same applies to the two lines below.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala [hunk 17]
```diff
@@ -29,165 +28,95 @@ import org.apache.spark.sql.types._
...
+    val newXMk = xMk + dx * (dx - dxN)
+    val newYMk = yMk + dy * (dy - dyN)
+
+    val isNull = Or(IsNull(x), IsNull(y))
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala]: Similar here. Is it the same as the following?

~~~scala
If(IsNull(x) || IsNull(y),
  Seq(count, xAvg, yAvg, ck, xMk, yMk),
  Seq(n, newXAvg, newYAvg, newCk, newXMk, newYMk))
~~~

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala [hunk 18]
```diff
@@ -29,165 +28,95 @@ import org.apache.spark.sql.types._
...
-
-  override def inputAggBufferAttributes: Seq[AttributeReference] = {
-    aggBufferAttributes.map(_.newInstance())
+  override val mergeExpressions: Seq[Expression] = {
+
+    val n1 = count.left
+    val n2 = count.right
+    val n = n1 + n2
+    val dx = xAvg.right - xAvg.left
+    val dxN = If(EqualTo(n, Literal(0.0)), Literal(0.0), dx / n)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Corr.scala]: `EqualTo` => `===`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Covariance.scala [hunk 19]
```diff
@@ -28,171 +26,90 @@ import org.apache.spark.sql.types._
...
+  override lazy val updateExpressions: Seq[Expression] = {
+    val n = count + Literal(1.0)
+    val dx = x - xAvg
+    val dy = y - yAvg
+    val dyN = dy / n
+    val newXAvg = xAvg + dx / n
+    val newYAvg = yAvg + dyN
+    val newCk = ck + dx * (dy - dyN)
+
+    val isNull = Or(IsNull(x), IsNull(y))
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Covariance.scala]: Ditto.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 20]
```diff
@@ -436,3 +436,21 @@ case class Murmur3Hash(children: Seq[Expression], seed: Int) extends Expression
     }
   }
 }
+
+/**
+  * An expression that will print the value of child to stderr (used for debugging codegen).
+  */
+case class Echo(child: Expression) extends UnaryExpression {
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I'm not sure whether `Echo` belongs to this PR. cc @nongli @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 21]
```diff
@@ -244,7 +244,7 @@ case class WholeStageCodegen(plan: CodegenSupport, children: Seq[SparkPlan])
 
     // try to compile, helpful for debug
     // println(s"${CodeFormatter.format(source)}")
-    CodeGenerator.compile(source)
+    // CodeGenerator.compile(source)
```
Comment (mengxr) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: Any reason to comment this line out?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 22]
```diff
@@ -198,7 +198,8 @@ case class Window(
           functions,
           ordinal,
           child.output,
-          (expressions, schema) => newMutableProjection(expressions, schema))
+          (expressions, schema) =>
+            newMutableProjection(expressions, schema, subexpressionEliminationEnabled))
```
Comment (mengxr) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: This would definitely help the computation of summary statistics. But would `subexpressionEliminationEnabled` have other side effects? cc @nongli

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala [hunk 23]
```diff
@@ -59,6 +59,55 @@ class BenchmarkWholeStageCodegen extends SparkFunSuite {
     benchmark.run()
   }
 
+  def testStddev(values: Int): Unit = {
```
Comment (mengxr) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: The test actually touches both `stddev` and `kurtosis`, which should be reflected by the name.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala [hunk 24]
```diff
@@ -59,6 +59,55 @@ class BenchmarkWholeStageCodegen extends SparkFunSuite {
...
+    }
+
+    benchmark.addCase("kurtosis w codegen") { iter =>
+      sqlContext.setConf("spark.sql.codegen.wholeStage", "true")
+      sqlContext.range(values).groupBy().agg("id" -> "kurtosis").collect()
+    }
+
+
+    /**
+    Using ImperativeAggregate:
```
Comment (mengxr) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: Add ` as implemented in Spark 1.6`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala [hunk 25]
```diff
@@ -148,7 +197,8 @@ class BenchmarkWholeStageCodegen extends SparkFunSuite {
   }
 
   test("benchmark") {
```
Comment (mengxr) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: Shall we add a comment and say those benchmarks are skipped in normal build, or change `test("benchmark")` to `ignore("benchmark")` to be clear.

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala [hunk 26]
```diff
@@ -325,6 +325,10 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     "drop_partitions_ignore_protection",
     "protectmode",
 
+    // Difference accuracy (we are better)
```
Comment (mengxr) [inline sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: Could you elaborate? For example, we can say Hive implementation has numerical stability issues.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/WindowQuerySuite.scala (hunk 3 above)
Comment (mengxr) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/WindowQuerySuite.scala]: It would be nice to see we can tolerate some small numerical differences in query tests. But this is out of scope here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala (hunk 11 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: These expression require `m3.right`, it is not valid if momentOrder < 3, will add a comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 27]
```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
 
-  override val aggBufferAttributes: Seq[AttributeReference] = Seq.tabulate(bufferSize) { i =>
-    AttributeReference(s"M$i", DoubleType)()
+    Seq(
+      /* count = */ n,
+      /* avg = */ newAvg,
+      /* m2 = */ newM2,
+      /* m3 = */ newM3,
+      /* m4 = */ newM4
+    ).take(momentOrder + 1)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: This could be simplified to `trimHigherOrder(Seq(n, newAvg, newM2, newM3, newM4))`, which won't reduce readability.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala (hunk 11 above)
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Then shall we make `updateExpression` follow the same style?

**Comment by mengxr** [comment]
@davies I made one pass. It would be nice to have a JIRA for checking query result with tolerance on numerical differences, because the result might change (though unlikely) if we merge the partial results in a different order.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 20 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: This is helpful to debug generated expression, especially for these complicated functions.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala (hunk 21 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: no reason, will bring it back.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 20 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: how is this used, or how can this be used?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala (hunk 11 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: `updateExpression` does not need that.

Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: I understand that `updateExpression` is not necessary. This is just to make the code more consistent. For example, I figured it out why this is not necessary in `updateExpression`, then I was confused by why we use the `if` branches here. The logic would be clearer if we use `if` branches in both methods.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala (hunk 8 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: The `if` here is not necessary, will be removed (compiler can remove it)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala (hunk 12 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: We may still want to have a good name when you call `explain`, I'd like to keep them.

Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Sounds good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 20 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: This could be used to print out the the result of any expression, for example
```
  Sqrt(Echo(a * b))
```
You can see the result of `a * b`

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: So this is only used if you construct the expressions manually?

I'd rename this to PrintToStderr. Echo means it is returning whatever you are typing in, which is not what this is about.

**Comment by davies** [comment]
@mengxr Thanks for reviewing this, I should had addressed all your comments.

**Comment by mengxr** [comment]
test this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 28]
```diff
@@ -52,178 +50,162 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
+  protected val m2 = AttributeReference("m2", DoubleType, nullable = false)()
+  protected val m3 = AttributeReference("m3", DoubleType, nullable = false)()
+  protected val m4 = AttributeReference("m4", DoubleType, nullable = false)()
+
+  private def trimHigherOrder[T](expressions: Seq[T]) = expressions.take(momentOrder + 1)
+
+  override val aggBufferAttributes = trimHigherOrder(Seq(n, avg, m2, m3, m4))
+
+  override val initialValues: Seq[Expression] =
+    trimHigherOrder(Seq(Literal(0.0), Literal(0.0), Literal(0.0), Literal(0.0), Literal(0.0)))
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: Sorry for some miscommunication. The previous inline comments are useful here because `Lit(0.0)` carries no information. The comments are not necessary when the variable names can clearly tell what they are. Please recover the inline comments for initial values.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala [hunk 29]
```diff
@@ -52,178 +50,162 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
...
-  override def checkInputDataTypes(): TypeCheckResult =
-    TypeUtils.checkForNumericExpr(child.dataType, s"function $prettyName")
+  override val mergeExpressions: Seq[Expression] = {
 
-  override def aggBufferSchema: StructType = StructType.fromAttributes(aggBufferAttributes)
+    val n1 = n.left
+    val n2 = n.right
+    val newN = n1 + n2
+    val delta = avg.right - avg.left
+    val deltaN = If(EqualTo(newN, Literal(0.0)), Literal(0.0), delta / newN)
```
Comment (mengxr) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: `EqualTo` => `===` (please check other occurrences)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala (hunk 28 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala]: There is no difference for these initial values, the order does not matter here. Do you still think we should keep those comments? Or should I change to use `fill()`?

**Comment by mengxr** [comment]
LGTM pending Jenkins. It is great to see 5x speedup!

**Comment by davies** [comment]
Merging this into master, thanks!
